### ISSUE TITLE
Rework span encoding: table decomposition + ragged embedding_bag

### ISSUE THREAD

**Issue by eiennohito** [ku-nlp/jumanpp#160]
## Problem

The current span encoding in `compute_loss` does a per-span matmul (512→64) for every positive and negative candidate. With 30 negatives × 8 span lengths × ~500 positions × batch size, this produces ~3.9M matmuls per forward pass — roughly 10× the cost of the transformer itself.

Additionally, the loss computation loops over span lengths k=1..8 in Python, processing each k as a separate tensor. This structure was imposed by the shared memory data layout (per-k `neg_k{k}` arrays), not by the math.

## Core insight

The span encoder `span_enc_proj(concat(embed(c₁), ..., embed(c₈)))` decomposes as:

```
encode(c₁, …, cₖ) = Σₛ Wₛ · E[cₛ]
```

This is a sum of per-slot linear transforms of codepoint embeddings. Precompute a combined table, then span encoding becomes a gather + sum — exactly what `F.embedding_bag(mode='sum')` does.

### Combined table (one matmul per forward pass)

```python
E_batch = self.embed.weight[batch_vocab]           # (U, dim) — gather batch-local
table = (E_batch @ self.span_enc_proj.weight.T)    # (U, 8*dim) — one matmul
table = table.view(U * 8, dim)                     # zero-copy view
```

C++ deduplicates codepoints per batch (U ≈ 3K vs V ≈ 20K). The matmul shrinks ~6×.

### Slot-offset encoding (C++)

Each span codepoint encoded as `local_id * max_span_len + slot` (uint32). Span encoding becomes a single `embedding_bag` lookup — no per-span matmul, no slot loop.

### Ragged layout

All spans (positive as candidate 0, then negatives) for all (batch, position, k) packed into one flat array. Bag sizes = span lengths (1,1,...,2,2,...,...,8,8,...). Eliminates per-k tensor separation.

Per-candidate entries: 1+2+...+8 = 36 (ragged) vs 8×8 = 64 (padded). ~44% less data per candidate set.

### Unified loss (no per-k loop)

```python
all_embs = F.embedding_bag(table, span_ids, offsets, mode='sum')
all_embs = all_embs.reshape(B, Lk, K, n_cand, dim)

# N/A substitution on positives
is_na = morph_len[:, 1:1+Lk].unsqueeze(-1) < k_vals
all_embs[:, :, :, 0] = torch.where(is_na.unsqueeze(-1), na_emb, all_embs[:, :, :, 0])

# scores + NCE — all vectorized
all_scores = torch.einsum('blkd,blknd->blkn', scoring, all_embs)
```

### Step 1: Validate decomposition (Python only)

Test that table gather+sum produces identical results to `encode_spans`. Verify `embedding_bag` autograd works through non-leaf weight. Catch PyTorch gotchas before touching C++.

### Step 2: New `compute_loss` with compatibility shim

Rewrite model.py to use table + embedding_bag internally. Accept old per-k data format via a shim that converts in Python (dedup + encode). Run training, verify loss/metrics match.

Files: `lm/training/tflm/model.py`

### Step 3: C++ batch server — new shared memory format

- Workers write neg spans to per-worker scratch buffers (local memory)
- After barrier, single-threaded post-pass:
  1. Scan all windows + neg buffers → dedup map (`uint16 local_id[MAX_VOCAB]`)
  2. Encode positive spans (from window text) as candidate 0 + negatives as candidates 1..n_neg, all as `local_id * 8 + slot` (uint32)
  3. Pack ragged into flat `span_ids` array
  4. Write `batch_vocab`, `n_unique` scalar

New slot layout:
```
windows:      uint16[B, seq_len]
morph_len:    uint8[B, seq_len]
n_unique:     uint32[1]
batch_vocab:  uint32[vocab_size]       — fixed allocation, first U entries valid
span_ids:     uint32[total_indices]    — ragged, fixed total count
```

`total_indices = B × Lk × n_cand × K(K+1)/2` is constant.

Files: `lm/preproc/batch_server.h`, `lm/preproc/batch_server.cc`

### Step 4: Python data consumer

Read new format. Precompute static tensors once (slot_indices pattern, bag_offsets — constant for the run). Return `(codepoints, morph_len, batch_vocab, span_ids)`.

Files: `lm/training/tflm/data.py`

### Step 5: Remove shim, final model.py

Drop the compatibility layer. `compute_loss` takes pre-encoded data directly.

Files: `lm/training/tflm/model.py`, `lm/training/train.py`

### Step 6: Update docs

`docs/transformer-lm-design.md`, `lm/CLAUDE.md`

## Things to verify

- `embedding_bag` autograd through non-leaf weight (step 1 catches this)
- Position validity mask: positions valid for k=1 but not k=8 need masking in the loss (~7 edge positions)
- Convergence parity: run training with old vs new, compare loss curves
